src/lib/hasher/solana.ts

```diff
@@ -0,0 +1,85 @@
+import { sha256 } from '@noble/hashes/sha256'
+import { PublicKey } from '@solana/web3.js'
+import { concat, zeroPadValue } from 'ethers'
+import { type CCIPMessage, type CCIPVersion, defaultAbiCoder } from '../types.js'
+import { type LeafHasher, LEAF_DOMAIN_SEPARATOR } from './common.js'
+
+export const getV16SolanaLeafHasher =
+  (
+    sourceChainSelector: bigint,
+    destChainSelector: bigint,
+    onRamp: string,
+  ): LeafHasher<CCIPVersion.V1_6> =>
+  (message: CCIPMessage<CCIPVersion.V1_6>): string =>
+    hashSolanaMessage(message, hashSolanaMetadata(sourceChainSelector, destChainSelector, onRamp))
+
+const encode = (target: string, value: string | bigint | number) =>
+  defaultAbiCoder.encode([target], [value])
+
+const encodeSolanaAddress = (address: string): Uint8Array => {
+  try {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
```

Comment:
`PublicKey` throws an error with missing type, I haven't found an easy way to address it with a type-safe wrapper.